core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2252,6 +2252,20 @@ object ZIO extends ZIOCompanionPlatformSpecific {
   )(f: PartialFunction[A, U]): ZIO[R, E, List[U]] =
     ZIO.collectAllParN(n)(in).map(_.collect(f))
 
+  /**
+   * Evaluate each effect in the structure from left to right, collecting the
+   * the successful values and discarding the empty cases. For a parallel version, see `collectSomePar`.
+   */
+  def collectSome[R, E, A, B](in: Iterable[A])(f: A => ZIO[R, Option[E], B]): ZIO[R, E, List[B]] =
+    foreach(in)(f).optional.map(_.fold(List.empty[B])(ZIO.identityFn))
```

Comment:
I don't think this is going to do what you want it to. Right now any missing value is going to cause `foreach` to fail. To get your intended semantics you need to do something like `foreach(in)(a => f(a).optional).map(_.flatten)`. Would be good to add some tests to verify the correct behavior here.